src/track/beats.h

```diff
@@ -1,177 +1,239 @@
+    FramePos getFirstBeatPosition() const;
+    FramePos getLastBeatPosition() const;
+    Beat getBeatAtIndex(int index) const;
+    void setAsDownbeat(int beatIndex);
+
+  private:
+    void updateBpm();
+    void scaleDouble();
+    void scaleTriple();
+    void scaleQuadruple();
+    void scaleHalve();
+    void scaleThird();
+    void scaleFourth();
+    void scaleMultiple(uint multiple);
+    void scaleFraction(uint fraction);
+    void generateBeatsFromMarkers();
+    void clearMarkers();
+    SINT getSampleRate() const;
+    double getDurationSeconds() const;
+    void setFirstBeatFrame(FramePos framePos);
+    FramePos getFirstBeatFrame() const;
+
+    QString m_subVersion;
+    Bpm m_bpm;
+    BeatList m_beats;
+    track::io::Beats m_beatsProto;
+    audio::StreamInfo m_streamInfo;
+    friend QDebug operator<<(QDebug dbg, const BeatsInternal& arg);
 };
 
-// Beats is a pure abstract base class for BPM and beat management classes. It
-// provides a specification of all methods a beat-manager class must provide, as
-// well as a capability model for representing optional features.
-class Beats : public QObject {
+/// Beats is a class for BPM and beat management classes.
+/// It stores beats information including beats position, down beats position,
+/// phrase beat position and changes in tempo.
+class Beats final : public QObject {
     Q_OBJECT
+  private:
+    explicit Beats(const audio::StreamInfo& streamInfo = audio::StreamInfo(),
+            const BeatsInternal& internal = BeatsInternal());
+
   public:
-    Beats() { }
     ~Beats() override = default;
+    Beats(const Beats& other) = delete;
+
+    /// The source of this byte array is the serialized representation of beats
+    /// generated by the protocol buffer and stored in the database.
+    void initWithProtobuf(const QByteArray& byteArray);
+    /// A list of beat locations in audio frames may be provided.
+    /// The source of this data is the analyzer.
+    void initWithAnalyzer(const QVector<FramePos>& beats,
+            const QVector<track::io::TimeSignatureMarker>&
+                    timeSignatureMarkers =
+                            QVector<track::io::TimeSignatureMarker>());
+
+    /// Serializes into a protobuf.
+    QByteArray toProtobuf() const;
+
+    /// Returns a string representing the version of the beat-processing code that
+    /// produced this Beats instance. Used by BeatsFactory for associating a
+    /// given serialization with the version that produced it.
+    QString getVersion() const;
+
+    /// Return a string that represent the preferences used to generate
+    /// the beats object.
+    QString getSubVersion() const;
+
+    void setSubVersion(const QString& subVersion);
+
+    /// Initializes the BeatGrid to have a BPM of dBpm and the first beat offset
+    /// of firstBeatFrame. Does not generate an updated() signal, since it is
+    /// meant for initialization.
+    void setGrid(Bpm dBpm, FramePos firstBeatFrame = kStartFramePos);
```

Comment:
Should this still exist?

Reply:
By removing this function, the alternative would be three steps:
1. clear()
2. setBpm(bpm, index=0)
3. setFirstBeatFrame(framePos)